command/013_config_upgrade.go

```diff
@@ -138,64 +470,79 @@ necessary adjustments, and then commit.
-	path := filepath.Join(dir, "providers.tf")
-	if _, err := os.Stat(path); !os.IsNotExist(err) {
-		for i := 1; ; i++ {
-			path = filepath.Join(dir, fmt.Sprintf("providers-%d.tf", i))
-			if _, err := os.Stat(path); os.IsNotExist(err) {
-				break
+	for name, rp := range requiredProviders {
+		// If there's already an explicit source, skip it
+		if rp.Source != "" {
+			continue
+		}
+
+		// Construct a legacy provider FQN using the existing addr's type. This
+		// is necessary because the config parser for required providers
+		// constructs a default provider FQN for configurations with no source.
+		// For this tool specifically we want to treat those as legacy
+		// providers, so that we can look up the namespace on the registry.
+		addr := addrs.NewLegacyProvider(rp.Type.Type)
+		p, err := getproviders.LookupLegacyProvider(addr, source)
+		if err == nil {
+			rp.Type = p
+		} else {
+			if _, ok := err.(getproviders.ErrProviderNotKnown); ok {
+				// Setting the provider address to a zero value struct
+				// indicates that there is no known FQN for this provider,
+				// which will cause us to write an explanatory comment in the
+				// HCL output advising the user what to do about this.
+				rp.Type = addrs.Provider{}
 			}
+			diags = diags.Append(tfdiags.Sourceless(
+				tfdiags.Warning,
+				"Could not detect provider source",
+				fmt.Sprintf("Error looking up provider source for %q: %s", name, err),
+			))
 		}
 	}
 
-	f, err := os.Create(path)
-	if err != nil {
-		diags = diags.Append(tfdiags.Sourceless(
-			tfdiags.Error,
-			"Unable to create providers file",
-			fmt.Sprintf("Error when generating providers configuration at '%s': %s", path, err),
-		))
-		return diags
-	}
-	defer f.Close()
+	return diags
+}
 
-	err = providersTemplate.Execute(f, providerSources)
-	if err != nil {
-		diags = diags.Append(tfdiags.Sourceless(
-			tfdiags.Error,
-			"Unable to create providers file",
-			fmt.Sprintf("Error when generating providers configuration at '%s': %s", path, err),
-		))
-		return diags
+// Take a list of tokens and a separator token, and return two lists: one up to
+// and including the first instance of the separator, and the rest of the
+// tokens. If the separator is not present, return the entire list in the first
+// return value.
+func partitionTokensAfter(tokens hclwrite.Tokens, separator hclsyntax.TokenType) (hclwrite.Tokens, hclwrite.Tokens) {
+	for i := 0; i < len(tokens); i++ {
+		if tokens[i].Type == separator {
+			return tokens[0 : i+1], tokens[i+1:]
+		}
 	}
 
-	return nil
+	return tokens, nil
+}
+
+// Generate a list of tokens for a comment explaining that a provider source
+// could not be detected.
+func noSourceDetectedComment(name string) hclwrite.Tokens {
+	comment := fmt.Sprintf(`# TF-UPGRADE-TODO
+#
+# No source detected for this provider. You must add a source address
+# in the following format:
+#
+# source = "your.domain.com/organization/%s"
```

Comment:
This is a nit and not super important, but... I try to use [explicitly reserved domains](https://en.wikipedia.org/wiki/Example.com) like `example.com` for things like this just in case we'd otherwise somehow cause grief for the legitimate owners of other domains were users to just copy-paste the example and not actually change it. The risk of grief seems low here, but perhaps `your-registry.example.com` gets across the same information without that (admittedly very low) risk.